tfjs-converter/src/executor/graph_executor.ts

```diff
@@ -380,19 +382,23 @@ export class GraphExecutor implements FunctionExecutor {
   private checkTensorForDisposalWithNodeLiveUntilInfo(
       node: Node, tensorMap: NamedTensorsMap, context: ExecutionContext,
       tensorsToKeep: Set<number>, outputNodeNameSet: Set<string>,
-      liveUntilNodes?: string[]) {
-    // Skip output nodes and any control flow nodes, since its dependency is
-    // tricky to track correctly.
-    if (isControlFlow(node) || outputNodeNameSet.has(node.name)) {
-      return;
+      liveUntilNodes?: Node[]) {
+    function isNonDisposableNode(node: Node) {
+      // Skip output nodes and any control flow nodes, since its dependency is
+      // tricky to track correctly.
+      return isControlFlow(node) || outputNodeNameSet.has(node.name);
     }
-    if (liveUntilNodes == null) {
+
+    if (isControlFlow(node) || liveUntilNodes == null) {
```

Comment:
I see you moved the check for node disposability to check the liveUntilNodes (this makes sense since those are the ones we're disposing). Do we need this `isControlFlow(node)` check? That means we'll never dispose any nodes that live until a control flow node if I'm not mistaken.

Is this to make sure we always have a control flow node's inputs available (e.g. in the case of a while)?

Reply:
> Is this to make sure we always have a control flow node's inputs available (e.g. in the case of a while)?
Yes. That is the behavior of the old algorithm. The branch here check whether the current node allow us to dispose the input of it, while the checks in the loop are for whether we should dispose the input itself. See https://github.com/tensorflow/tfjs/pull/7505/files#diff-44e0a825cd7c6f31c03d9333db5dc21a8937d66a5b28d719c699863eef96ad8dR329

But the behavior is changed. The old function does not allow disposing inputs of an output node no matter it's control flow or not, while this PR allows it. I don't see any current or future problems can have with it.


> Do we need this isControlFlow(node) check? 
No, bc a graph with control flow should not be sync. But I still keep it in case the upstream graph builder has some special cases.